File: LeetCode/680.py
class Solution:
    def remove_one(self, s: str) -> (str, str):
        l = len(s)
        for i in range(l):
            i, j = i, l - 1 - i
            if i == j: return None, None
            if s[i] != s[j]:
                return s[:i] + s[i+1:], s[:j] + s[j+1:]
        return None, None
                
    def check_pal(self, s) -> bool:
        l = len(s)
        if l % 2:
            return s[:l//2] == s[l//2+1:][::-1]
        return s[:l//2] == s[l//2:][::-1]
    
    def validPalindrome(self, s: str) -> bool:
        if len(s) <= 1: return True
        s1, s2 = self.remove_one(s)
        print(s1, s2)
        if s1 and s2:
            return self.check_pal(s1) or self.check_pal(s2)
        return True

File: LeetCode/680.cc
class Solution {
    bool validPalindrome(int i, int j, const string& s, bool deleted) {
        while(i < j) {
            if(s[i] == s[j]) i++, j--;
            else if(deleted) return false;
            else return validPalindrome(i+1, j, s, true) || validPalindrome(i, j-1, s, true);
        }
        return true;
    }
    
public:
    bool validPalindrome(string s) {
        const int n = s.length();
        return validPalindrome(0, n-1, s, false);
    }
};